### ISSUE TITLE
Which functions to call for a batch-export?

### ISSUE THREAD

**Issue by VitusW42** [KhronosGroup/glTF-Blender-IO#1161]
Hello again!

As I wrote before (https://github.com/KhronosGroup/glTF-Blender-IO/issues/1135), I forked the Khronos exporter and extended it by a few custom functionalities. The way I did this was to copy and modify _gltf2_blender_export.py_ and _gltf2_blender_gltf2_exporter.py_. From within those files, I am calling the functions and classes from the default Blender gltf Exporter. This works very well, but due to the rather complicated structure of the exporter, I'm having some trouble with the next step:

Another function I want to include in my Blender addon is the ability of a batch export of the scene. The goal is to generate multiple gltf+bin files for different LODs of the object. For this, the exporter is looking for keywords in the collections, so all objects of an LOD collection will be exported separate from other objects in other collections.

The main exporter function is the save(context, export_settings) in _gltf2_blender_export.py_. And I started sorting out my collections in that method. Where I'm stuck now is how I can limit the objects to export, to those in the lod collections.

I hope someone here can give me some pointers how to do this sort of batch export.

**Comment by scurest** [comment]
I don't believe there is any easy way to do this right now. Related issues:

#192 Export specific collection  
#1073 support context override

**Comment by VitusW42** [comment]
Hi, thanks for helping out again!
It seems like the "selected objects" export works fine now. One cheeky solution would probably be to make a selection of the first collection, run the exporter with "selected objects" active, then select the next collection and export and so on.
It's just not a very elegant way to do this and I was hoping there's a more straight-forward to this by calling some functions and passing values manually.
I'm fine with the way being hard, I already forked the exporter to make some further changes, I just can't figure out in which function the exporter is actually getting the scene's data from.

**Comment by scurest** [comment]
Gathering the glTF data starts in [`gather_gltf2`](https://github.com/KhronosGroup/glTF-Blender-IO/blob/92a2e5d24a0de25242f6c879c13fd6a1426ce581/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather.py#L27), which calls `__gather_scene`, which calls `gather_node` on the root objects to start gathering nodes.

**Comment by VitusW42** [comment]
Would it be possible to instantiate a new, empty, scene object and for each batch, copy the objects from the collection into this new scene - then pass the new scene to the __gather_gltf2 function?

**Comment by scurest** [comment]
Can't you just, like, replace `blender_scene` here 

https://github.com/KhronosGroup/glTF-Blender-IO/blob/92a2e5d24a0de25242f6c879c13fd6a1426ce581/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather.py#L54

with the collection you want to export from?

**Comment by VitusW42** [comment]
I started doing that, but I keep running into issues with it. I set up a simple scene with three collections with one primitive in each of them. When I run my exporter, it generates the three gltf file, but only the first file actually contains data and only one .bin file is being generated.
While it is almost certain that I am doing things wrong, it feels to me as if the exporter doesn't quite lend itself to what I'm trying to do: 
It seems that there is multiple places in multiple scripts in which the exporter is fetching data directly from the scene and it's writing that data into the files. Hence my idea to create an instance of a scene, fill it with the relevant objects and just pass that :)

**Comment by VitusW42** [comment]
I got this figured out now and I'll provide some of the insights I gained from this here. Maybe it'll help some other people who want to try a similar thing.

Instead of pursuing one of the possible solution I described above, I looked further into what @scurest described. I had my exporter working to the extend that it would create the gltf files, according to the LODs, but it wouldn't fill it with any geometry information. I wrongly concluded that it must've something to do with the scene object, hence my suggestion to create a new one. But the answer is much simpler:
the Khronos exporter was caching the result of the _gather_scene function. And so when my exporter called the function a second time, it'd skip the step.

**The solution is simply to remove the `@cached ` property of the function**, of course after modifying the function in a way scurest describes above.

So thank you again for putting me on the right track, you've been incredibly helpful!

I would provide my code, but it is very customized to our application, so I don't think it'd add more value to what has already been said above.
